### PR TITLE
feat(access): review/enforce consistent API access controls

### PR DESCRIPTION
**Pull request by david-leifker**
* PolicyConfig - Privileges organized by API Group and Operation
* AuthUtils/AuthenticationUtil/ControllerUtil/ResourceUtils - consolidated into AuthUtil
* Added EXISTS_ENTIY privilege
* Added CREATE_ENTITY privilege (CreateIfNotExistsValidator)
* Added CreateIfNotExists flag to OpenAPI create endpoints
* Deprecated SEARCH privilege
* Entity Indices systemCreated timestamp
* Fix cache key calculation
* Fix search view privileges inheritance
* Fix MeResolver for viewing self

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Review by chriscollins3456 — COMMENTED** [review]
things are looking solid here!

I also pulled down locally and i think there's a few things that could cause a negative user experience.

1. The `listRecommendations` endpoint will often throw an error if we try and recommend anything that you don't have access to. Maybe you've thought about this already or maybe we could filter suggestions out on the java side?
2.  If we add that permissions check on `ContainerType` and a user doesn't have permissions to a parent container but does have permissions on a child dataset, then the search query will fail when returning that dataset and going to the entity profile page will fail. We need to consider how to handle this...
3. Similarly, if you have access to data flows but not child data jobs then we raise an error on entity profile page when we try and hydrate these child jobs. maybe there are more scenarios like this?

Overall this does lock things down much more, but it could result in a clunky experience that requires some fine tuning.

**Comment by chriscollins3456** [comment]
also to extend my original comment - we run into the same issue if an entity has a domain/glossary term on it and the user doesn't have permission to see it. Then we raise an error in search when it shows up and on the profile page.

maybe we include a default permission that all users can view all glossary terms/nodes, domains, and data products?

Same idea applies for users and groups. I know we're not checking permissions for them now and the more I think about it the more I think that makes sense?

Or if we want to allow orgs to make that call for themselves, have a default permission allowing users to see all of them by default (basically, you can see all of the datahub entities by default, and anything we get from ingestion will be opt in permission)

**Comment by chriscollins3456** [comment]
Or we could modify these resolvers further and filter out glossary term/domain/dataproduct/owners if they don't have permission to view one. But that would be a large chunk of work to be honest.

**Comment by chriscollins3456** [comment]
final thought (for now) maybe we could only optionally apply this permission check on that hydration layer if we know they're querying for this entity directly through the API instead of just hydrating it as a reference on another entity?

(1) I'm not sure how we could do that but we could probably find a way
(2) Would this be acceptable from a security standpoint? Basically, you can't query an entity directly, but if it's a subquery on another entity you do have access to do we say that you can get info about it by it being connected to the entity you have access to?

this could solve the woes I was describing in the previous comments if we think it makes sense and is secure enough. And that initial layer of the API would always be safe, you could only get info on entities you don't have access to view if you have access to view a parent entity in the query

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/authorization/AuthorizationUtils.java [hunk 1]
```diff
@@ -191,37 +194,32 @@ public static boolean canDeleteQuery(
...
+      throw new AuthorizationException("Unauthorized to view this entity");
+    }
   }
 
   public static boolean isAuthorized(
       @Nonnull QueryContext context,
-      @Nonnull Optional<EntitySpec> resourceSpec,
-      @Nonnull PoliciesConfig.Privilege privilege) {
+      @Nullable EntitySpec resourceSpec,
+      @Nonnull final Disjunctive<Conjunctive<PoliciesConfig.Privilege>> privileges) {
```
[the remark this answers is not in the document]
Comment (david-leifker) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/authorization/AuthorizationUtils.java:210]: I'll create the simple function here and hook up it up anywhere in the graphql core where it only requires a single privilege.

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/notebook/NotebookType.java [hunk 2]
```diff
@@ -171,6 +171,15 @@ public List<DataFetcherResult<Notebook>> batchLoad(
...
     final List<Urn> urns = urnStrs.stream().map(UrnUtils::getUrn).collect(Collectors.toList());
     try {
+      // if search authorization is disabled, skip the view permission check
+      if (context
+          .getOperationContext()
+          .getOperationContextConfig()
+          .getSearchAuthorizationConfiguration()
+          .isEnabled()) {
+        AuthorizationUtils.checkViewPermissions(urns, context);
+      }
```
[the remark this answers is not in the document]
Comment (david-leifker) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/notebook/NotebookType.java:181]: If I add this, then lots of the UI will break unless we explicitly enable for common policies to say allow dataset entities. @chriscollins3456 - lets chat about this.

**Review by RyanHolstien — APPROVED** [review]
Overall LGTM, some comments/nits

File: metadata-utils/src/main/java/com/linkedin/metadata/authorization/PoliciesConfig.java [hunk 3]
```diff
@@ -624,17 +628,236 @@ public class PoliciesConfig {
...
+    @Nonnull private String type;
     private String displayName;
     private String description;
 
     static Privilege of(String type, String displayName, String description) {
       return new Privilege(type, displayName, description);
     }
+
+    @Override
+    public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (david-leifker) [reply metadata-utils/src/main/java/com/linkedin/metadata/authorization/PoliciesConfig.java:848]: I am actually being explicit here. Will add a comment.

[the remark this answers is not in the document]
Comment (david-leifker) [reply metadata-utils/src/main/java/com/linkedin/metadata/authorization/PoliciesConfig.java:848]: I am explicitly ignoring displayName and description

File: metadata-service/restli-servlet-impl/src/main/java/com/linkedin/metadata/resources/operations/Utils.java [hunk 4]
```diff
@@ -35,17 +33,16 @@ public static String restoreIndices(
...
-    if (Boolean.parseBoolean(System.getenv(REST_API_AUTHORIZATION_ENABLED_ENV))
-        && !isAuthorized(
-            authentication,
+    if (!isAPIAuthorized(
+            AuthenticationContext.getAuthentication(),
             authorizer,
-            ImmutableList.of(PoliciesConfig.RESTORE_INDICES_PRIVILEGE),
+            Disjunctive.disjoint(PoliciesConfig.RESTORE_INDICES_PRIVILEGE),
             resourceSpec)) {
       throw new RestLiServiceException(
```
[the remark this answers is not in the document]
Comment (david-leifker) [reply metadata-service/restli-servlet-impl/src/main/java/com/linkedin/metadata/resources/operations/Utils.java:47]: Should be 403, nice catch!

**Pull request closed without merging**